tensorizer/serialization.py

```diff
@@ -4446,17 +4128,492 @@ def write_state_dict(self, state_dict: Union[Dict, List, Tuple]):
+
+        for w in cuda_specs:
+            assert w.tensor_data_task is None
+            w.tensor_data_task = cuda_transfer.submit(w)
+            self._jobs.append(w.tensor_data_task)
+
+        return cuda_transfer.executor
+
+    def _prepare_for_write_contiguous(self, write_specs: Sequence[_WriteSpec]):
+        def make_contiguous(write_spec, dependency):
+            if dependency is not None:
+                dependency.result(_TIMEOUT)
+            write_spec.tensor = write_spec.tensor.contiguous()
+            write_spec.data_length = write_spec.tensor.nbytes
+            write_spec.user_owns_tensor_data = False
+
+        for w in write_specs:
+            # if there is a tensor_data_task it is a cuda tensor
+            if w.tensor_data_task is not None or w.tensor.is_contiguous():
+                continue
+            w.tensor_data_task = self._computation_pool.submit(
+                make_contiguous, w, w.tensor_data_task
+            )
+
+    def _prepare_for_write_dtype(self, write_specs: Sequence[_WriteSpec]):
+        torch_dtype_to_numpy_dtype_cache: Dict[str, str] = {}
+
+        for w in write_specs:
+            tensor_dtype_str = str(w.tensor.dtype)
+            if _NumpyTensor._is_asymmetric(w.tensor.dtype):
+                # is opaque
+                w.dtype = (
+                    f"<V{w.tensor.element_size():d}"
+                    + OPAQUE_DTYPE_SEP
+                    + tensor_dtype_str
+                )
+                w.set_min_file_version_number(OPAQUE_TENSORIZER_VERSION)
+            else:
+                w.dtype = torch_dtype_to_numpy_dtype_cache.get(
+                    tensor_dtype_str, None
+                )
+                if w.dtype is None:
+                    torch_dtype_to_numpy_dtype_cache[tensor_dtype_str] = (
+                        w.dtype
+                    ) = (
+                        torch.tensor((), dtype=w.tensor.dtype, device="cpu")
+                        .numpy()
+                        .dtype.str
+                    )
+
+    @staticmethod
+    def _do_clone(write_spec, dependency: Optional[_Future]):
+        if dependency is not None:
+            dependency.result(_TIMEOUT)
+        write_spec.tensor = write_spec.tensor.clone().detach()
+        write_spec.numpy_tensor = _NumpyTensor.from_tensor(write_spec.tensor)
+
+    def _prepare_for_write_encryption(
+        self, write_specs: Sequence[_WriteSpec]
+    ) -> None:
+        assert self._encrypted and self._encryption is not None
+
+        # If any tensors are shared, so we need to clone all but one of them before encrypting
+        write_specs_by_addr: Dict[int, List[TensorSerializer._WriteSpec]] = (
+            collections.defaultdict(list)
+        )
+        for w in write_specs:
+            if w.tensor.device.type != "cpu":
+                continue
+            address = w.tensor.untyped_storage().data_ptr()
+            write_specs_by_addr[address].append(w)
```

Comment:
Interestingly, PyTorch counts tensors imported from other libraries like `numpy` as always having separate storages, even if they overlap or represent a completely identical region when compared with other storages—even in its `Tensor.is_set_to` function, so non-PyTorch-native tensors are sort of a "known bug" for this method, though it's niche enough to not affect almost any normal workflows.

Other than that, we can identify overlapping input data more precisely using an $O \lparen n \cdot \log \lparen n \rparen \rparen$ (where $n$ is the number of tensors) algorithm I designed for it once, since this old way incorrectly identifies disjoint slices of a tensor as being shared, but we can add that in another PR, since it's not a regression of this one.

Reply:
https://github.com/coreweave/tensorizer/issues/146